libvirt/domain_def.go

```diff
@@ -1,216 +1,76 @@
-	Type     string `xml:"type,attr,omitempty"`
-	File     string `xml:",chardata"`
-}
-
-// <nvram>/var/lib/libvirt/qemu/nvram/sled12sp1_VARS.fd</nvram>
-type defNvRam struct {
-	File string `xml:",chardata"`
-}
-
-type defController struct {
-	Type  string `xml:"type,attr,omitempty"`
-	Model string `xml:"model,attr,omitempty"`
-}
 
-type defConsole struct {
-	Type   string `xml:"type,attr"`
-	Source struct {
-		Path string `xml:"path,attr,omitempty"`
-	} `xml:"source"`
-	Target struct {
-		Type string `xml:"type,attr,omitempty"`
-		Port string `xml:"port,attr"`
-	} `xml:"target"`
-}
-
-type defFilesystemReadOnly bool
-
-type defFilesystem struct {
-	Type       string                `xml:"type,attr"`
-	AccessMode string                `xml:"accessmode,attr"`
-	ReadOnly   defFilesystemReadOnly `xml:"readonly",omitempty`
-	Source     struct {
-		Dir string `xml:"dir,attr,omitempty"`
-	} `xml:"source"`
-	Target struct {
-		Dir string `xml:"dir,attr,omitempty"`
-	} `xml:"target"`
-}
-
-// The filesystem element has a <readonly/> tag when
-// the host directory cannot be written by the guest. When
-// the <readonly/> tag is omitted the read-write permissions
-// are granted.
-// To show the empty <readonly/> tag we have to define a
-// "alias" of bool and provide a custom marshaller for it.
-func (r defFilesystemReadOnly) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
-	if !r {
-		return nil
-	}
-	err := e.EncodeElement("", start)
-	return err
-}
+	"github.com/libvirt/libvirt-go-xml"
+)
 
-func (r *defFilesystemReadOnly) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
-	var value string
-	err := d.DecodeElement(&value, &start)
-	if err != nil {
-		return err
+func newFilesystemDef() libvirtxml.DomainFilesystem {
+	return libvirtxml.DomainFilesystem{
+		Type:       "mount",  // This is the only type used by qemu/kvm
+		AccessMode: "mapped", // A safe default value
+		ReadOnly:   &libvirtxml.DomainFilesystemReadOnly{},
 	}
-
-	*r = value == ""
-
-	return nil
-}
-
-func newFilesystemDef() defFilesystem {
-	fs := defFilesystem{}
-
-	// This is the only type used by qemu/kvm
-	fs.Type = "mount"
-
-	// A safe default value
-	fs.AccessMode = "mapped"
-	fs.ReadOnly = true
-
-	return fs
 }
 
 // Creates a domain definition with the defaults
 // the provider uses
-func newDomainDef() defDomain {
-	// libvirt domain definition
-	domainDef := defDomain{}
+func newDomainDef() libvirtxml.Domain {
+	domainDef := libvirtxml.Domain{
+		OS: &libvirtxml.DomainOS{
+			Type: &libvirtxml.DomainOSType{
+				Type: "hvm",
+			},
+		},
+		Memory: &libvirtxml.DomainMemory{
+			Unit:  "MiB",
+			Value: 512,
+		},
+		VCPU: &libvirtxml.DomainVCPU{
+			Placement: "static",
+			Value:     1,
+		},
+		CPU:     &libvirtxml.DomainCPU{},
+		Devices: &libvirtxml.DomainDeviceList{},
+		Features: &libvirtxml.DomainFeatureList{
+			PAE:  &libvirtxml.DomainFeature{},
+			ACPI: &libvirtxml.DomainFeature{},
+			APIC: &libvirtxml.DomainFeatureAPIC{},
```

Comment:
AFAIK we added `ACPI` and `APIC` at the beginning of the project because they were needed to properly shut down a domain. Without the it looked like Windows 95's "you can now turn off your computer" :)